### PR TITLE
Aida 818

### PR DESCRIPTION
**Pull request by sharknc**
This PR contains the contents of four separate tickets, AIDA-760, AIDA-762, AIDA-763, and AIDA-818. This should be treated as four separate code reviews.

AIDA-760
- Developed batch validator initialization script in main/docker/scripts/main.py.
- Verifies / validates all environment variables at startup
- SNS message is sent to the specified topic notifying users that the validation process has begun for the S3 submission archive
- Verifies s3 submission archive format to be .zip / .tar.gz / .tgz
- Downloads / extracts archive from S3 into local staging directory
- Uploads all ttl files to S3 validation bucket and queues S3 object path on SQS
- Creates a log of all files that have been sent to the sqs queue for future validation
- Enters indefinite loop to monitor the status of all AWS batch worker jobs until processing has completed.

AIDA-762
- Developed batch worker script in worker/docker/scripts/worker.py
- Verifies / validates all environment variables at startup
- Waits for SQS queue to be created based on AWS batch job id
- Consumes SQS messages and runs AIF validation on each in a subprocess
- Captures all AIF validator output and places it in the appropriate S3 prefix directory
- Captures all stdout output from AIF Validator and creates a new log file which is uploaded to the appropriate S3 directory
- Checks for the sourcefile.done and if the queue is empty before determining that there are no more messages to process and exits.

AIDA-763
- Developed batch finalization script in main/docker/scripts/main.py.
- After processing has completed in all workers the script will sync the S3 validation bucket.
- The contents of the synced bucket are verified against the sourcefiles.done log file to ensure that all messages that were sent to SQS queue are accounted for in the resulting synced s3 bucket directory
- A new verification file is created with the verification results and added to the synced s3 bucket directory
- A new tar in the format of {s3_submission_archive name}-result.tar.gz is created with the contents of the sync s3 bucket directory and uploaded back to the s3 validation bucket
- An SNS message is sent to the specified topic that the validation process has completed

AIDA-818
- Updated main.py and worker.py by creating Main and Worker python classes. 
- Updated functions to be private class functions of the Main and Worker classes
- Updated to use a single session of boto3 instead of creating new connections on each request
- Created a metric calculation function that is used in generating more informative final SNS message 
- Created new debug environment variables in main.py to set timeout and sleep intervals
- Updated all Docker run.sh.example scripts to account for changes.

### CODE REVIEW

File: docker/worker/scripts/worker.py [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+		"""
+		file_name = Path(file_path).name
+
+		try:
+			cmd = self.validation_home + '/target/appassembler/bin/validateAIF ' + self.validation_flags + ' -f '
+			cmd += file_path
+
+			logging.info("Executing AIF Validation for file %s with flags %s", file_name, self.validation_flags)
+			#**********************
+			# Requies python 3.7+ *
```
Comment (jakevb) [inline docker/worker/scripts/worker.py]: Typo: Requies should be Requires.

File: docker/worker/scripts/worker.py [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+		each to the specified bucket and location in S3.
+
+		:param str validation_dir: The validation directory that contains the output
+		:param str s3_object_prefix: The s3 prefix that will be appended to the uploaded files
+		:param str extension: The extension to search on in the validation directory
+		"""
+		items = glob.glob(validation_dir + '**/*' + extension)
+
+		if len(items) <= 0: 
+			logging.error("No validation ouptut files found in validation folder with extension %s", extension)
```
Comment (jakevb) [inline docker/worker/scripts/worker.py]: Typo: ouptut should be output.

File: docker/worker/scripts/worker.py [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+				self.node_index, self.queue_init_timeout)
+
+	
+	def _bucket_exists(self):
+		"""Helper function that will check if a validation bucket
+		exists.
+
+		:returns: True if bucket exists, False otherwise
+		:raises ClientError: S3 resource exception
+		:rasses ValueError: The validation bucket does not exist
```
Comment (jakevb) [inline docker/worker/scripts/worker.py]: Typo: rasses should be raises.

File: docker/worker/run.sh.example [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	-e VALIDATION_TIMEOUT='' \
+	-e VALIDATION_HOME='' \
+	-e VALIDATION_FLAGS= '' \
+        -e S3_VALIDATION_BUCKET='' \
+        -e AWS_BATCH_JOB_ID='' \
+        -e AWS_BATCH_JOB_NODE_INDEX= \
+        -e AWS_ACCESS_KEY_ID='' \
+        -e AWS_SECRET_ACCESS_KEY='' \
+        -e AWS_DEFAULT_REGION='' \
+        --name batch-worker \
```
Comment (jakevb) [inline docker/worker/run.sh.example]: Same as the note in the main/run.sh.  Also, AWS_BATCH_JOB_NODE_INDEX does not have ' ' marks.

File: docker/worker/README.md [hunk 5]
```diff
@@ -25,9 +25,9 @@ Save, restart Docker via `service docker restart` and confirm experimental is en
 $ docker version -f '{{.Server.Experimental}}'
 ```
 
-## Building AIF Validator Docker image
+## Building Batch Worker Docker image
 
-To build the docker image, copy the `build.sh.example` file provided to a new file, `build.sh`. This script is made up of a docker build command with various `--build-arg` build arguments. These arguments should be configured to meet the needs of your particular AIF Validator Docker image. Each argument with an example value is described in the table below. 
+To build the docker image, copy the `build.sh.example` file provided to a new file, `build.sh`. This script is made up of a docker build command with various `--build-arg` build arguments. These arguments should be configured to meet the needs of your particular Batch Worker Docker image. Each argument with an example value is described in the table below. 
 
 | Build Arg          |Example Value       | Drescription  |
```
Comment (jakevb) [inline docker/worker/README.md]: Typo: Drescritpion should be Description.

File: docker/main/scripts/main.py [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+        return False
+
+    for k, v in envs.items():
+        if not is_env_set(k, v):
+            return False
+
+    #check if master sleep interval can be converted to int
+    try:
+        int(envs['MAIN_SLEEP_INTERVAL'])
+    except InitilizationError:
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: InitilizationError should probably be changed to InitializationError.

File: docker/main/scripts/main.py [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+        :raises Exception: Excpetion occured when attempting to write to file 
+        """
+        file_path = '/'.join([results_path, source_verfication_path])
+        try:
+            with open(file_path, "w") as f:
+                print(message, file=f)
+                self.verification = message #set for final report
+        except:
+            logging.error("Error when writing source verification file %s", file_path)
+
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo Line 601: reuslts should be results.
Typo Line 600 and 610: You might want to change source_verfication_path to source_verification_path.

File: docker/main/scripts/main.py [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+            try: 
+                with open(queued_log) as file:
+                    sqs_objects = [Path(line.strip()).name for line in file]
+            except :
+                logging.error("Exception occured when reading %s during verification of validation", queued_log)
+
+                self._create_verification_output(
+                    results_path,
+                    verification_log,
+                    "Exception occured when reading {0} during verification of validation".format(queued_log)
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo Lines 546 and 551: occured should be occurred.

File: docker/main/scripts/main.py [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+            logging.error("Error [%s] occured when syncing s3 bucket %s with prefix %s", 
+                str(e.returncode), self.bucket, self.job_id)
+
+
+    def _verify_validation(self, results_path):
+        """Verifies that all files enqueued on SQS were accounted for in the 
+        resulting S3 bucket after validation. 
+
+        :param str results_path: The local path of the sync'd s3 bucket
+        :retruns: True if all files are account for, False otherwise
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: retruns should be returns.

File: docker/main/scripts/main.py [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+            logging.info("Syncing S3 bucket %s with prefix %s", self.bucket, self.job_id)
+            #**********************
+            # Requies python 3.7+ *
+            #**********************
+            output = subprocess.run(cmd, check=True, shell=True)
+            logging.info("Succesfully downloaded all files from s3 bucket %s with prefix %s", 
+                self.bucket, self.job_id)
+            
+        except CalledProcessError as e:
+            logging.error("Error [%s] occured when syncing s3 bucket %s with prefix %s",
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: occured should be occurred.

File: docker/main/scripts/main.py [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+        :param str dest_path: The local destination path where files will be syned to
+        :raises CalledProcessError: Subprocess exception when executing aws cli sync
+            command
+        """
+        try:
+            cmd = 'aws s3 sync s3://' + self.bucket + '/' + self.job_id + ' ' + dest_path
+
+            logging.info("Syncing S3 bucket %s with prefix %s", self.bucket, self.job_id)
+            #**********************
+            # Requies python 3.7+ *
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: Requies should be Requires.

File: docker/main/scripts/main.py [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+            #**********************
+            output = subprocess.run(cmd, check=True, shell=True)
+            logging.info("Succesfully downloaded all files from s3 bucket %s with prefix %s",
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: Succesfully should be Successfully.

File: docker/main/scripts/main.py [hunk 13]
```diff
@@ -0,0 +1,823 @@
...
+        path = Path(self.submission)
+        stem = Path(path.with_suffix('')).stem
+        return stem
+
+
+    def _sync_s3_bucket(self, dest_path):
+        """Helper function that will sync the s3 validation bucket with job id prefix to
+        the current working directory.
+
+        :param str dest_path: The local destination path where files will be syned to
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: syned should be synced.

File: docker/main/scripts/main.py [hunk 14]
```diff
@@ -0,0 +1,823 @@
...
+                response = batch_client.list_jobs(
+                    multiNodeJobId=self.job_id
+                )
+
+                job_list = response['jobSummaryList']
+                logging.info("AWS Batch job summary list %s", job_list)
+                running_jobs = list(filter(lambda job: job['status'] == 'RUNNING', job_list))
+
+
+                # check if no jobs are running, throw an error becasue master should still be running
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: becasue should be because.

File: docker/main/scripts/main.py [hunk 15]
```diff
@@ -0,0 +1,823 @@
...
+
+        except ClientError as e:
+            logging.error(e)
+
+
+    def _debug_wait_for_processing(self):
+        """This function is used for debugging purposes. This will remove any depencency on 
+        AWS batch and allow for jobs to be processe for the specified amount of time set in 
+        the [processing_timeout] parameter. Once that amount of time has elapsed this function will
+        return true.
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typos: depencency should be dependency and processe should be processed

File: docker/main/scripts/main.py [hunk 16]
```diff
@@ -0,0 +1,823 @@
...
+                    with open(self.source_log, 'a+') as f:
+                        f.write(s3_object + '\n')
+
+                    # upload file to S3
+                    self._upload_file_to_s3(self.source_log, self.job_id)
+
+                else:
+                    logging.error("Unable to add %s as SQS message", s3_object)
+            
+            # append .done to the sourceifles path
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: sourceifles should be sourcefiles.

File: docker/main/scripts/main.py [hunk 17]
```diff
@@ -0,0 +1,823 @@
...
+        s3_client = self.session.client('s3')
+
+        try:
+            for filepath in Path(self.job_id).glob('**/*.ttl'):
+                s3_object = '/'.join([self.job_id, 'UNPROCESSED', filepath.name])
+
+                logging.info("Uploading %s to bucket %s", s3_object, self.bucket)
+                s3_client.upload_file(str(filepath), self.bucket, s3_object)
+
+                # add the mssage to SQS
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: mssage should be message.

File: docker/main/scripts/main.py [hunk 18]
```diff
@@ -0,0 +1,823 @@
...
+        self._delete_sqs_queue(queue_url)
+
+
+    def _bucket_exists(self):
+        """Helper function that will check if a validation bucket
+        exists.
+
+        :returns: True if bucket exists, False otherwise
+        :raises ClientError: S3 resource exception
+        :rasses ValueError: The validation bucket does not exist
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo: rasses should be raises.

File: docker/main/scripts/main.py [hunk 19]
```diff
@@ -0,0 +1,823 @@
...
+        self._enqueue_files(queue_url)
+
+        # check for debug mode
+        if self.debug:
+            self._debug_wait_for_processing()
+        else:
+            # wait for all AWS batch jobs to complete processing
+            self._wait_for_processing()
+
+        # download all validation files from s3 for the currnet job
```
Comment (jakevb) [inline docker/main/scripts/main.py]: Typo:  currnet should be current.

File: docker/main/run.sh.example [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+
+docker run --rm -it \
+        -e S3_SUBMISSION_ARCHIVE='' \
+        -e S3_VALIDATION_BUCKET='' \
+        -e AWS_BATCH_JOB_ID='' \
+        -e AWS_BATCH_JOB_NODE_INDEX= '' \
+        -e MAIN_SLEEP_INTERVAL= '' \
+	-e WORKER_INIT_TIMEOUT= '' \
+	-e AWS_SNS_TOPIC_ARN= '' \
+	-e DEBUG= '' \
```
Comment (jakevb) [inline docker/main/run.sh.example]: You might want to remove the extra space between the = and the ' for AWS_BATCH_JOB_NODE_INDEX and other values that have that extra space.

File: docker/main/scripts/main.py (hunk 7 above)
Comment (sharknc) [reply docker/main/scripts/main.py]: Not sure I understand your second comment. Is there a typo in there?

Comment (jakevb) [reply docker/main/scripts/main.py]: The variable name source_verfication_path has verification spelled incorrectly.  It isn't necessarily important and the variable when referenced is also spelled incorrectly, so it doesn't effect any code.  I just felt that I should mention it.

Comment (sharknc) [reply docker/main/scripts/main.py]: ah okay, thanks!

**Pull request merged**